In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## Import the data and understand it

In [2]:
data = pd.read_csv('/Users/haoli/Downloads/Project_16_data.tsv', sep = '\t')

In [3]:
data.head()

rating       date         variation  \
0       5  31-Jul-18  Charcoal Fabric    
1       5  31-Jul-18  Charcoal Fabric    
2       4  31-Jul-18    Walnut Finish    
3       5  31-Jul-18  Charcoal Fabric    
4       5  31-Jul-18  Charcoal Fabric    

                                    verified_reviews  feedback  
0                                      Love my Echo!         1  
1                                          Loved it!         1  
2  Sometimes while playing a game, you can answer...         1  
3  I have had a lot of fun with this thing. My 4 ...         1  
4                                              Music         1

In [4]:
data.shape

(3150, 5)

In [5]:
data.rating.value_counts()

5    2286
4     455
1     161
3     152
2      96
Name: rating, dtype: int64

In [6]:
data.feedback.value_counts()

1    2893
0     257
Name: feedback, dtype: int64

In [7]:
data.variation.value_counts()

Black  Dot                      516
Charcoal Fabric                 430
Configuration: Fire TV Stick    350
Black  Plus                     270
Black  Show                     265
Black                           261
Black  Spot                     241
White  Dot                      184
Heather Gray Fabric             157
White  Spot                     109
White                            91
Sandstone Fabric                 90
White  Show                      85
White  Plus                      78
Oak Finish                       14
Walnut Finish                     9
Name: variation, dtype: int64

In [8]:
data.groupby('rating').describe()

feedback                                   
          count mean  std  min  25%  50%  75%  max
rating                                            
1         161.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
2          96.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
3         152.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0
4         455.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0
5        2286.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0

In [9]:
data[data['rating'] == 1].head(10)

rating       date          variation  \
141       1  30-Jul-18   Charcoal Fabric    
162       1  30-Jul-18  Sandstone Fabric    
341       1  28-Jul-18   Charcoal Fabric    
350       1  31-Jul-18              Black   
361       1  29-Jul-18              Black   
368       1  28-Jul-18              Black   
369       1  28-Jul-18              Black   
373       1  27-Jul-18              Black   
374       1  26-Jul-18              Black   
381       1  25-Jul-18              White   

                                      verified_reviews  feedback  
141                                 Not much features.         0  
162  Stopped working after 2 weeks ,didn't follow c...         0  
341                             Alexa hardly came on..         0  
350  Item no longer works after just 5 months of us...         0  
361  This thing barely works. You have to select 3r...         0  
368  I returned 2 Echo Dots & am only getting refun...         0  
369                                        not working         0  
373          I'm an Echo fan but this one did not work         0  
374                                                            0  
381  It worked for a month or so then it stopped. I...         0

In [10]:
data.date.value_counts()

30-Jul-18    1603
29-Jul-18     482
28-Jul-18     250
26-Jul-18      85
31-Jul-18      81
             ... 
21-May-18       2
9-Jun-18        1
4-Jun-18        1
13-Jun-18       1
29-May-18       1
Name: date, Length: 77, dtype: int64

In [11]:
data_0 = data[data['feedback'] == 0]
data_1 = data[data['feedback'] == 1]

In [12]:
data_1.shape[0]

2893

In [13]:
data_0.shape[0]

257

In [14]:
data_1_sample = data_1.sample(257)

In [15]:
data = pd.concat([data_1_sample, data_0])

In [18]:
data.head()

rating       date                     variation  \
2911       5  30-Jul-18                    White  Dot   
866        5  30-Jul-18          Heather Gray Fabric    
2228       5  30-Jul-18  Configuration: Fire TV Stick   
2454       5  31-Jul-18                    Black  Dot   
1611       3  29-Jul-18                   White  Show   

                                       verified_reviews  feedback  
2911                  Love the echo dot it’s amaxing!!!         1  
866       Super easy set up and am loving our new Echo!         1  
2228  I'm amazed at how many movies, tv shows and mo...         1  
2454                                   Family loved it!         1  
1611  I bought it to go with my security camera syst...         1

In [28]:
data.reset_index(inplace = True)

In [29]:
data.head()

index  rating       date                     variation  \
0   2911       5  30-Jul-18                    White  Dot   
1    866       5  30-Jul-18          Heather Gray Fabric    
2   2228       5  30-Jul-18  Configuration: Fire TV Stick   
3   2454       5  31-Jul-18                    Black  Dot   
4   1611       3  29-Jul-18                   White  Show   

                                    verified_reviews  feedback  
0                  Love the echo dot it’s amaxing!!!         1  
1      Super easy set up and am loving our new Echo!         1  
2  I'm amazed at how many movies, tv shows and mo...         1  
3                                   Family loved it!         1  
4  I bought it to go with my security camera syst...         1

## Preprocessing the text

In [26]:
import re
import nltk
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
from nltk.tokenize import word_tokenize

In [30]:
STOPWORDS = set(stopwords.words('english'))
corpus=[]
for i in range(514):
    review = re.sub('[^a-zA-Z]', ' ', data['verified_reviews'][i])
    review = review.lower()
    review = review.split()
    stemmer = PorterStemmer()
    review = [stemmer.stem(token) for token in review if not token in STOPWORDS]
    #contain all words that are not in stopwords dictionary
    review=' '.join(review)
    corpus.append(review)
corpus

['love echo dot amax',
 'super easi set love new echo',
 'amaz mani movi tv show avail plan drop cabl servic soon happi know lot view pleasur fire stick pay outrag cabl bill',
 'famili love',
 'bought go secur camera system perform like thought would alert us someon door also talk person door echo show cell phone',
 'work great',
 'enjoy still learn command ran one frustrat issu user friendli tri set voic recognit household',
 'echo great home voic rang ask question say command turn light favorit play music though hous',
 'awesom devic home full use inform us e phone answer spur moment question',
 'without cellphon cannot use mani featur ipad see use great alarm u r almost deaf hear alarm bedroom live room reason enough keep fun ask random question hear respons seem smartbon polit yet',
 'listen well way expect love',
 'easi set easi use',
 'around week love control item hous even make feel better away hous peopl know gone alexa disappoint',
 'easi set work like suppos',
 'got gift coo

In [31]:
#Explaination of the code:
#First, stopwords is a list, STOPWORDS is a hashset, when we check the existence of an element, hashset is faster;
#re.sub replace [^a-zA-Z] by ' '.
#Make all the letters to the lower case, and split the sentences into words.
#Use Portstemmer to normalize the words, for example change 'loved' to 'love'.
#' '.join(review) connects the normalized words using ' '.

In [32]:
from tqdm._tqdm_notebook import tqdm_notebook
tqdm_notebook.pandas()

/Users/haoli/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.*` instead of `tqdm._tqdm_notebook.*`
  """Entry point for launching an IPython kernel.
/Users/haoli/opt/anaconda3/lib/python3.7/site-packages/tqdm/std.py:658: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version
  from pandas import Panel


In [33]:
all_words = [word for text in tqdm_notebook(corpus) for word in text.split()]

In [34]:
all_words

['love',
 'echo',
 'dot',
 'amax',
 'super',
 'easi',
 'set',
 'love',
 'new',
 'echo',
 'amaz',
 'mani',
 'movi',
 'tv',
 'show',
 'avail',
 'plan',
 'drop',
 'cabl',
 'servic',
 'soon',
 'happi',
 'know',
 'lot',
 'view',
 'pleasur',
 'fire',
 'stick',
 'pay',
 'outrag',
 'cabl',
 'bill',
 'famili',
 'love',
 'bought',
 'go',
 'secur',
 'camera',
 'system',
 'perform',
 'like',
 'thought',
 'would',
 'alert',
 'us',
 'someon',
 'door',
 'also',
 'talk',
 'person',
 'door',
 'echo',
 'show',
 'cell',
 'phone',
 'work',
 'great',
 'enjoy',
 'still',
 'learn',
 'command',
 'ran',
 'one',
 'frustrat',
 'issu',
 'user',
 'friendli',
 'tri',
 'set',
 'voic',
 'recognit',
 'household',
 'echo',
 'great',
 'home',
 'voic',
 'rang',
 'ask',
 'question',
 'say',
 'command',
 'turn',
 'light',
 'favorit',
 'play',
 'music',
 'though',
 'hous',
 'awesom',
 'devic',
 'home',
 'full',
 'use',
 'inform',
 'us',
 'e',
 'phone',
 'answer',
 'spur',
 'moment',
 'question',
 'without',
 'cellphon',
 'c

In [35]:
voca = nltk.FreqDist(all_words)

In [36]:
voca

FreqDist({'echo': 159, 'work': 157, 'use': 109, 'devic': 107, 'alexa': 105, 'music': 93, 'love': 90, 'like': 89, 'great': 87, 'amazon': 74, ...})

In [37]:
voca['love']

90

In [38]:
len(voca)

1476

In [39]:
voca_lst = [word for word in voca]

In [40]:
voca_lst[:10]

['love', 'echo', 'dot', 'amax', 'super', 'easi', 'set', 'new', 'amaz', 'mani']

## Word embedding using Tfidf

In [41]:
from sklearn.feature_extraction.text import TfidfVectorizer 

In [42]:
tf_vec = TfidfVectorizer(vocabulary = voca_lst, ngram_range=(1, 2), min_df=2)

In [43]:
X = tf_vec.fit_transform(corpus)

In [44]:
y = data['feedback']

In [45]:
from sklearn.model_selection import train_test_split

In [46]:
x_train, x_test, y_train, y_test = train_test_split(X.toarray(), y, random_state=42)

In [47]:
x_train

array([[0.        , 0.24151768, 0.30188625, ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.12787868, 0.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.13788633, 0.17235173, ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [48]:
y

0      1
1      1
2      1
3      1
4      1
      ..
509    0
510    0
511    0
512    0
513    0
Name: feedback, Length: 514, dtype: int64

## Random Forest Tree

In [49]:
from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics

In [50]:
rtf3 = RandomForestClassifier(n_estimators=200, max_features="auto")

In [51]:
rtf3.fit(x_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=200,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [52]:
print(metrics.classification_report(y_test, rtf3.predict(x_test)))

              precision    recall  f1-score   support

           0       0.70      0.88      0.78        60
           1       0.87      0.67      0.75        69

    accuracy                           0.77       129
   macro avg       0.78      0.77      0.77       129
weighted avg       0.79      0.77      0.77       129



In [54]:
rtf5 = RandomForestClassifier(max_depth = 5, n_estimators=200, max_features="auto")

In [55]:
rtf5.fit(x_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=5, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=200,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [56]:
print(metrics.classification_report(y_test, rtf5.predict(x_test)))

              precision    recall  f1-score   support

           0       0.70      0.70      0.70        60
           1       0.74      0.74      0.74        69

    accuracy                           0.72       129
   macro avg       0.72      0.72      0.72       129
weighted avg       0.72      0.72      0.72       129



## DNN with dropout layers (MLP) 

In [57]:
from tensorflow.keras.models import Model, Sequential
from tensorflow.keras.layers import Input, Dense, Dropout
from tensorflow.keras.optimizers import Adam

In [59]:
def build_mlp_model(input_dim, layers, output_dim, dropout_rate=0.2):
    X = Input(shape=(input_dim,))
    
    H = X
    for layer in layers:
        H = Dense(layer, activation='relu')(H)
        H = Dropout(rate=dropout_rate)(H) 
        
    activation_func = 'sigmoid'
    
    Y = Dense(output_dim, activation=activation_func)(H)
    return Model(inputs=X, outputs=Y)

In [74]:
mlp_model = build_mlp_model(input_dim = 1476,
                           layers = [700, 350, 100],
                           output_dim = 1,
                           dropout_rate = 0.3)

In [75]:
mlp_model.compile(loss = 'binary_crossentropy', optimizer = Adam(), metrics = ['accuracy'])

In [76]:
amz_history = mlp_model.fit(
    x_train, 
    y_train, 
    epochs=20, 
    shuffle=True, 
    validation_data=(x_test, y_test), )

Train on 385 samples, validate on 129 samples
Epoch 1/20
385/385 [==============================] - 0s 1ms/sample - loss: 0.6900 - accuracy: 0.5766 - val_loss: 0.6775 - val_accuracy: 0.8140
Epoch 2/20
385/385 [==============================] - 0s 251us/sample - loss: 0.6397 - accuracy: 0.8597 - val_loss: 0.6012 - val_accuracy: 0.8450
Epoch 3/20
385/385 [==============================] - 0s 231us/sample - loss: 0.4188 - accuracy: 0.9143 - val_loss: 0.4196 - val_accuracy: 0.8217
Epoch 4/20
385/385 [==============================] - 0s 220us/sample - loss: 0.1359 - accuracy: 0.9662 - val_loss: 0.4919 - val_accuracy: 0.7829
Epoch 5/20
385/385 [==============================] - 0s 223us/sample - loss: 0.0617 - accuracy: 0.9740 - val_loss: 0.5603 - val_accuracy: 0.8295
Epoch 6/20
385/385 [==============================] - 0s 227us/sample - loss: 0.0457 - accuracy: 0.9818 - val_loss: 0.6257 - val_accuracy: 0.8295
Epoch 7/20
385/385 [==============================] - 0s 231us/sample - loss: 0.

#### We see that the accuracy can be improved to 0.8450. 